### PR TITLE
Space Calculator Structure

### PR DESCRIPTION
**Pull request by jimleeder123**
This is what we discussed yesterday (05/02/2024) and is base don the diagram that was drawn. The strucutre has been set up for the calculator with no real data in it. The only "real" data perhaps is where a `SpaceCalculatorInput` model can be transformed into an `Inputs` class. This will all be expanded upon in future PRs / cards. I have left some questions in comments on the files.

### CODE REVIEW

**Comment by jimleeder123** [comment]
[Space calculator structure](https://trello.com/c/RJa8CSSC/56-space-calculator-structure)

File: app/Services/SpaceCalculator/Inputs.php [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+        $this->growthPercentage = $growthPercentage;
+        $this->deskPercentage = $deskPercentage;
+        $this->hybridWorking = $hybridWorking;
+        $this->mobility = $mobility;
+        $this->collaboration = $collaboration;
+
+        /* todo: discuss - should we be passing in enums or strings?
+        Depends how much of a "black box" this should be... (The enums are defined in the Laravel app)
+        */
+    }
```
Comment (gbuckingham89) [inline app/Services/SpaceCalculator/Inputs.php:65]: We can simplify this by using "[constructor property promotion](https://platform.sh/blog/php-80-feature-focus-constructor-property-promotion/)" - i.e. declaring the property accessors in the constructor, so we could replace all of the above with simply;

```php
    public function __construct(
        public readonly Workstyle $workstyle,
        public readonly int $totalPeople,
        public readonly int $growthPercentage,
        public readonly int $deskPercentage,
        public readonly HybridWorking $hybridWorking,
        public readonly Mobility $mobility,
        public readonly Collaboration $collaboration,
    ) {

    }
```

File: app/Services/SpaceCalculator/Inputs.php [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+        $this->workstyle = $workstyle;
+        $this->totalPeople = $totalPeople;
+        $this->growthPercentage = $growthPercentage;
+        $this->deskPercentage = $deskPercentage;
+        $this->hybridWorking = $hybridWorking;
+        $this->mobility = $mobility;
+        $this->collaboration = $collaboration;
+
+        /* todo: discuss - should we be passing in enums or strings?
+        Depends how much of a "black box" this should be... (The enums are defined in the Laravel app)
```
Comment (gbuckingham89) [inline app/Services/SpaceCalculator/Inputs.php:63]: Don't want to be using strings if we can avoid - they're not "strict" enough - nothing to stop you passing in an invalid option!

Good thinking on the black box approach though.

If you were doing it super-strict, you'd probably be building this calculator as a standalone package anyway, in which case you'd probably include the enums in it.

For us though, I think we're OK leaving the enums where they are for now - especially as enums are just simple PHP classes - they're not tied into Laravel itself at all.

File: app/Models/SpaceCalculatorInput.php [hunk 3]
```diff
@@ -59,4 +60,20 @@ public function enquiry(): BelongsTo
     {
         return $this->belongsTo(Enquiry::class, 'id', 'enquiry_id');
     }
+
+    /**
+     * @return Inputs
+     */
+    public function transformToInputs(): Inputs
```
Comment (gbuckingham89) [inline app/Models/SpaceCalculatorInput.php:67]: Maybe naming it `transformToCalculatorInputs()` would be clearer?

File: app/Models/SpaceCalculatorInput.php [hunk 4]
```diff
@@ -59,4 +60,20 @@ public function enquiry(): BelongsTo
...
+    {
+        return new Inputs(
+            $this->workstyle,
+            $this->total_people,
+            $this->growth_percentage,
+            $this->desk_percentage,
+            $this->hybrid_working,
+            $this->mobility,
+            $this->collaboration,
+        );
```
Comment (gbuckingham89) [inline app/Models/SpaceCalculatorInput.php:77]: _Not nesecarirly something to change, more of any FYI._

I know it's pretty obvious in this case as we've named them well and they pretty much map 1-to-1, but in PHP 8 you can use [named arguments](https://stitcher.io/blog/php-8-named-arguments), so you don't have to remember which order the arguments should be passed in - and makes it clearer when reading;

```php
return new Inputs(
     workstyle: $this->workstyle,
     totalPeople: $this->total_people,
     growthPercentage: $this->growth_percentage,
     deskPercentage: $this->desk_percentage,
     hybridWorking: $this->hybrid_working,
     mobility: $this->mobility,
     collaboration: $this->collaboration,
);
```

It's particularly useful when calling the `run()` method on actions, as you don't get any auto-completion!

File: tests/Unit/Models/SpaceCalculatorInputTest.php [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+
+class SpaceCalculatorInputTest extends TestCase
+{
+    public function test_transform_to_inputs(): void
+    {
+        $model = SpaceCalculatorInput::factory()->create();
+
+        $inputs = $model->transformToInputs();
+
+        $this->assertInstanceOf(Inputs::class, $inputs);
```
Comment (gbuckingham89) [inline tests/Unit/Models/SpaceCalculatorInputTest.php:17]: We could maybe add assertions that we've mapped the properties correctly - e.g;

```php
$this->assertEquals($model->total_people, $inputs->totalPeople);
```

File: tests/Unit/Services/SpaceCalculator/CalculatorTest.php [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+namespace Tests\Unit\Services\SpaceCalculator;
+
+use App\Models\SpaceCalculatorInput;
+use App\Services\SpaceCalculator\Calculator;
+use App\Services\SpaceCalculator\Output;
+use Tests\TestCase;
+
+class CalculatorTest extends TestCase
+{
+    public function test_returns_output_class_from_calculate_method(): void
```
Comment (gbuckingham89) [inline tests/Unit/Services/SpaceCalculator/CalculatorTest.php:12]: Perfect for now 😍

**Review by gbuckingham89 — CHANGES_REQUESTED** [review]
This is great - exactly what we were aiming for - nice job! 👍

I've answered your questions and left a few ideas for simplifications / minor improvements.

File: app/Services/SpaceCalculator/Inputs.php (hunk 1 above)
Comment (jimleeder123) [reply app/Services/SpaceCalculator/Inputs.php:65]: Done, have added this to my list of things to read up on!

File: app/Services/SpaceCalculator/Inputs.php (hunk 2 above)
Comment (jimleeder123) [reply app/Services/SpaceCalculator/Inputs.php:63]: Ok thanks sounds good, will leave it as it is

File: app/Services/SpaceCalculator/Inputs.php (hunk 1 above)
Comment (gbuckingham89) [reply app/Services/SpaceCalculator/Inputs.php:65]: There was nothing wrong with what we had - but the newer method saves us some work / means less duplication!

File: app/Models/SpaceCalculatorInput.php (hunk 3 above)
Comment (jimleeder123) [reply app/Models/SpaceCalculatorInput.php:67]: Done

File: app/Models/SpaceCalculatorInput.php (hunk 4 above)
Comment (jimleeder123) [reply app/Models/SpaceCalculatorInput.php:77]: Done, and thanks another for the reading list!

File: tests/Unit/Models/SpaceCalculatorInputTest.php (hunk 5 above)
Comment (jimleeder123) [reply tests/Unit/Models/SpaceCalculatorInputTest.php:17]: Done for all properties

**Review by gbuckingham89 — APPROVED** [review]
Perfect! 🤩

**Pull request merged**